File: Solution/Piyush_Umate_task2_ModelBasedCF.py
import sys
from pyspark import SparkContext, SparkConf
from pyspark.mllib.recommendation import ALS, MatrixFactorizationModel, Rating
import time,math
from operator import add

file_path = sys.argv[1]
test_file = sys.argv[2]
output_file = "Piyush_Umate_ModelBasedCF.txt"
t = time.time()
def user_movie_map(line):
    l = line.split(',')
    if len(l) == 2:
        return (int(l[0]), int(l[1])), None
    return (int(l[0]),int(l[1])), float(l[2])

def compute_range(value):
    if value >= 0.0 and value < 1.0:
        return 0
    elif value >= 1.0 and value < 2.0:
        return 1
    elif value >= 2.0 and value < 3.0:
        return 2
    elif value >= 3.0 and value < 4.0:
        return 3
    else:
        return 4

spark_context = SparkContext(appName='ModelBasedCF', conf=SparkConf())
spark_context.setLogLevel("WARN")
test_file = spark_context.textFile(test_file)
# test_file = test_file.coalesce(4)
test_file_header = test_file.first()
testing_data = test_file \
    .filter(lambda line: line != test_file_header) \
    .map(lambda line: user_movie_map(line))\
    .persist()

# print "testing data", testing_data.count()

data = spark_context.textFile(file_path)
# data = data.coalesce(4)
header = data.first()
data = data \
    .filter(lambda line: line != header) \
    .map(lambda line: user_movie_map(line)) \
    .persist()


training_data = data.subtractByKey(testing_data).map(
    lambda r: Rating(r[0][0],r[0][1],r[1])
).persist()
# print training_data.count(), "training data"
#testing_data = testing_data.keys().persist()



rank = 10
numIterations = 12

model = ALS.train(training_data, rank, numIterations, 0.1)
predictions = model.predictAll(testing_data.keys()).map(lambda r: ((r[0], r[1]), r[2]))
ratesAndPreds =data.map(lambda r: r).join(predictions)
RMSE = math.sqrt(ratesAndPreds.map(lambda r: (r[1][0] - r[1][1])**2).mean())
# print predictions.collect()n

def toCSVLine(k, v):
    return str(k[0]) + ', ' + str(k[1]) + ', ' + str(v)


lines = predictions.sortByKey().map(lambda (k, v): toCSVLine(k, v)).collect()

file = open(output_file, 'w')
file_content = '\n'.join(lines)
file.write(file_content)

baseline = ratesAndPreds.map(
        lambda r: (compute_range(abs(r[1][0]-r[1][1])),1)
    ).countByKey()\
    .items()

for (index, count) in baseline:
    if index == 0:
        print(">=0 and <1: "+ str(count))
    elif index == 1:
        print(">=1 and <2: "+ str(count))
    elif index == 2:
        print(">=2 and <3: " + str(count))
    elif index == 3:
        print(">=3 and <4: "+ str(count))
    else:
        print(">=4: "+str(count))

print("RMSE: " + str(RMSE))
print("Time: %s sec" % str(time.time() - t))
# model.save(spark_context, "target")


#
# print data.collect()
# print testing_data.collect()
# print training_data.count()
# print time.time() - t

# l = spark_context.parallelize([("a", 1), ("b", 1), ("a", 1)])
# print l.countByKey().items()

File: Solution/Piyush_Umate_task2_ModelBasedCF.scala
import java.io.{File, PrintWriter}


import org.apache.spark.rdd.RDD
import org.apache.spark.{HashPartitioner, SparkConf, SparkContext}
import org.apache.spark.mllib.recommendation.{ALS, Rating}

import scala.collection.immutable.ListMap

object ModelBasedCF {
  val output_file = "Piyush_Umate_ModelBasedCF.txt"
  def user_movie_map(line: String): ((Int, Int), Double) = {
    var l = line.split(",")
    if (l.size == 2) {
      ((l(0).toInt, l(1).toInt), -1.0)
    } else {
      ((l(0).toInt, l(1).toInt), l(2).toDouble)
    }
  }

  def compute_range(value: Double): Int = {
    if (value >= 0.0 && value < 1.0) {
      0
    } else if(value >= 1.0 && value < 2.0) {
      1
    } else if(value >= 2.0 && value < 3.0) {
      2
    } else if(value >= 3.0 && value < 4.0) {
      3
    } else {
      4
    }
  }

  def main(args: Array[String]): Unit = {
    val t0 = System.currentTimeMillis()
    val conf = new SparkConf()
    val spark_context = new SparkContext(conf)
    spark_context.setLogLevel("WARN")
    val test_file = spark_context.textFile(args(1))
    val test_file_header = test_file.first()
    val testing_data = test_file.filter(line => line != test_file_header)
      .map(line => user_movie_map(line))
      .persist()

    var data_stringy = spark_context.textFile(args(0))
    val header = data_stringy.first()
    val data = data_stringy.filter(line => line != header)
        .map(line => user_movie_map(line))
        .persist()

    val training_data = data.subtractByKey(testing_data)
      .map(_ match {case ((user_id, movie_id), rating) => Rating(user_id, movie_id, rating)})
      .persist()

    val rank = 10
    val numIterations = 12
    val lambda = 0.1

    val model = ALS.train(
      training_data, rank, numIterations, lambda
    )

    val predictions = model.predict(testing_data.keys).map { case Rating(user, product, rate) =>
      ((user, product), rate)}.sortByKey()

    val ratesAndPreds = data.map(identity).join(predictions)

    val MSE = ratesAndPreds.map { case ((user, product), (r1, r2)) =>
      val err = (r1 - r2)
      err * err
    }.mean()

    val RMSE = math.sqrt(MSE)

//    println(RMSE)

    var baseline = ratesAndPreds.map{ case ((user, product), (r1, r2)) =>
      (compute_range(math.abs(r1 - r2)),1)}.countByKey()

    baseline = ListMap(baseline.toSeq.sortBy(_._1):_*)


    val predictions_dict = predictions.collect()
    val writer = new PrintWriter(new File(output_file))
    val last_index = predictions_dict.size - 1

    for ((((user_id, movie_id),rating), index) <- predictions_dict.zipWithIndex) {

      val formatted_string = Array(
        user_id.toString(),
        movie_id.toString(),
        rating.toString()
      ).mkString(", ")

      if (index != last_index) {
        writer.write(s"$formatted_string\n")
      } else {
        writer.write(s"$formatted_string")
      }
    }
    writer.close()

    for((k,v)<-baseline) {
      if(k == 0) {
        println(">=0 and <1: "+ v)
      } else if (k == 1) {
        println(">=1 and <2: "+ v)
      } else if(k == 2) {
        println(">=2 and <3: "+ v)
      } else if(k == 3) {
        println(">=3 and <4: "+ v)
      } else {
        println(">=4: "+ v)
      }
    }
//    println(baseline)

    println("RMSE: "+ RMSE)
    println("Time: " + (System.currentTimeMillis() - t0)/1000 + "sec")
  }
}
